lib/output/writer/kinesis.go

```diff
@@ -87,16 +113,82 @@ func NewKinesis(
 	if len(conf.PartitionKey) == 0 {
 		return nil, errors.New("partition key must not be empty")
 	}
-	return &Kinesis{
+
+	k := Kinesis{
 		conf:         conf,
 		log:          log.NewModule(".output.kinesis"),
 		stats:        stats,
 		hashKey:      text.NewInterpolatedString(conf.HashKey),
 		partitionKey: text.NewInterpolatedString(conf.PartitionKey),
-	}, nil
+		streamName:   aws.String(conf.Stream),
+	}
+
+	b := backoff.NewExponentialBackOff()
+	if conf.Backoff.InitialInterval != "" {
+		d, err := time.ParseDuration(conf.Backoff.InitialInterval)
+		if err != nil {
+			return nil, fmt.Errorf("invalid backoff initial interval: %v", err)
+		}
+		b.InitialInterval = d
+	}
+	if conf.Backoff.MaxInterval != "" {
+		d, err := time.ParseDuration(conf.Backoff.MaxInterval)
+		if err != nil {
+			return nil, fmt.Errorf("invalid backoff max interval: %v", err)
+		}
+		b.MaxInterval = d
+	}
+	if conf.Backoff.MaxElapsedTime != "" {
+		d, err := time.ParseDuration(conf.Backoff.MaxElapsedTime)
+		if err != nil {
+			return nil, fmt.Errorf("invalid backoff max elapsed interval: %v", err)
+		}
+		b.MaxElapsedTime = d
+	}
+	k.backoff = backoff.WithMaxRetries(b, conf.Backoff.MaxRetries)
+
+	return &k, nil
 }
 
-// Connect attempts to establish a connection to the target SQS queue.
+//------------------------------------------------------------------------------
+
+// toRecords converts an individual benthos message into a slice of Kinesis
+// batch put entries by promoting each message part into a single part message
+// and passing each new message through the partition and hash key interpolation
+// process, allowing the user to define the partition and hash key per message
+// part.
+func (a *Kinesis) toRecords(msg types.Message) ([]*kinesis.PutRecordsRequestEntry, error) {
+	entries := make([]*kinesis.PutRecordsRequestEntry, msg.Len())
+
+	err := msg.Iter(func(i int, p types.Part) error {
+		m := message.New(nil)
+		m.Append(p)
+
+		entry := kinesis.PutRecordsRequestEntry{
+			Data:         p.Get(),
+			PartitionKey: aws.String(a.partitionKey.Get(m)),
```

Comment:
I believe this is a breaking change from the previous implementation which applied the same partition key to all Kinesis records, would you prefer I change the logic here to be backwards compatible?

Reply:
Nah this is fine, it's a new target and I think this behaviour makes much more sense. In fact, I might change the Kafka output to do the same. As long as it's documented in the output description it should be fine.